File: Data Structures & Algorithms/eating-bananas/submission-2.py
class Solution:
    def count_hours(self, piles, k):
        return sum([math.ceil(i / k) for i in piles])

    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        n = len(piles)
        if h == n:
            return max(piles)
        r = max(piles)
        l = 1
        res = r
        while l <= r:
            middle = (r + l) // 2
            hours = self.count_hours(piles, middle)
            print(hours, middle)
            if hours <= h:
                res = middle
                r = middle - 1
            else:
                l = middle + 1
        return res

File: Data Structures & Algorithms/max-water-container/submission-2.java
class Solution {
    public int maxArea(int[] heights) {
        int res = -1;
        int l = 0;
        int r = heights.length - 1;
        while (l < r){
            int target = (r - l) * Math.min(heights[l], heights[r]);
            res = Math.max(res, target);
            if (heights[l] < heights[r])
                l += 1;
            else
                r -= 1;
        }
        return res;
    }
}
